### PR TITLE
Adding formatSPFileSize which formats file sizes into the exact same format as SharePoint document libraries

### PR DESCRIPTION
**Pull request by LeonArmston**
This pull request introduces a new Handlebars helper for formatting file sizes in a way that matches SharePoint's native display, along with its documentation and implementation. The main changes are the addition of the `formatSPFileSize` helper, its registration, and comprehensive documentation and examples.

image

**New Handlebars helper for SharePoint-style file size formatting:**

* Added a new helper function `helperFormatSPFileSize` in `HandlebarsHelpers.ts` to format byte values as human-readable file sizes using binary (1024-based) units, matching SharePoint’s display conventions. Includes logic for bytes, KB, MB, GB, and TB with tiered decimal precision.
* Registered the new `formatSPFileSize` helper in the exported helpers map so it can be used in Handlebars templates.

**Documentation:**

* Added documentation (**See Below**) for the new `formatSPFileSize` helper in `templating.md`, including usage instructions, a detailed example, and a table of sample inputs and outputs to clarify expected behavior.

#### formatSPFileSize

**Syntax** `{{formatSPFileSize <value>}}`

**Description** Formats a file size in bytes as a human-readable string that matches SharePoint’s native file size display. Before using this helper, map a *RefinableInt* managed property to `OWS_FILESIZEDISPLAY`, then map that property to a slot such as `fileSizeDisplay` and use that slot in the Handlebars expression. The helper uses binary (1024-based) divisors with tiered decimal precision.

**Example** `{{formatSPFileSize (slot item @root.slots.fileSizeDisplay)}}` (for example, if the slot value is `10485760`, it returns `10 MB`).

| Input (bytes) | Output |
|---|---|
| 0 | `0 bytes` |
| 500 | `500 bytes` |
| 1023 | `1023 bytes` |
| 1024 | `1 KB` |
| 1536 | `1.5 KB` |
| 34972 | `34.2 KB` |
| 51200 | `50 KB` |
| 148098 | `145 KB` |
| 1048576 | `1 MB` |
| 1209452 | `1.15 MB` |
| 10485760 | `10 MB` |
| 11624602 | `11.1 MB` |
| 82849344 | `79.0 MB` |
| 104857600 | `100 MB` |
| 1073741824 | `1.00 GB` |
| 8486862848 | `7.90 GB` |
| 1099511627776 | `1.00 TB` |

### CODE REVIEW

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 1]
```diff
@@ -232,6 +232,48 @@ function helperHexToString(hex: any): string {
...
+    if (mb < 100) return `${mb.toFixed(1)} MB`;
+    return `${Math.round(mb)} MB`;
+}
+
+function formatGigabytes(gb: number): string {
+    if (Number.isInteger(gb)) return `${Math.round(gb)} GB`;
+    if (gb < 10) return `${gb.toFixed(2)} GB`;
+    if (gb < 100) return `${gb.toFixed(1)} GB`;
+    return `${Math.round(gb)} GB`;
+}
```
[the remark this answers is not in the document]
Comment (wobba) [reply search-parts/src/helpers/HandlebarsHelpers.ts]: @LeonArmston Make the change?

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 2]
```diff
@@ -232,6 +232,48 @@ function helperHexToString(hex: any): string {
...
+
+function formatTerabytes(tb: number): string {
+    if (Number.isInteger(tb)) return `${Math.round(tb)} TB`;
+    return `${tb.toFixed(1)} TB`;
+}
```
[the remark this answers is not in the document]
Comment (wobba) [reply search-parts/src/helpers/HandlebarsHelpers.ts]: @LeonArmston  same

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 3]
```diff
@@ -232,6 +232,48 @@ function helperHexToString(hex: any): string {
...
+
+function helperFormatSPFileSize(bytes: any): Handlebars.SafeString {
+    const b = Number.parseInt(String(bytes), 10);
+    if (Number.isNaN(b) || b < 0) return new Handlebars.SafeString("0 KB");
+
```
[the remark this answers is not in the document]
Comment (wobba) [reply search-parts/src/helpers/HandlebarsHelpers.ts]: @LeonArmston you decice

[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts]: I've made a change - when I upload a empty file to SP it displays as nothing i.e. ""

File: search-parts/src/helpers/HandlebarsHelpers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts]: No I have done a different change. As seems in a SP doc lib you can get it to display 1 GB and 1.00 GB or 1 MB or 1.00MB

I produced some test file edge cases around 1GB and 1MB and it produced diff values depending on how close to the GB or MB.

Here is my GB working in the doc l;ib

image

Here is the GB working and showing the same values in Search:

image

The same logic applies for MB

**Comment by LeonArmston** [comment]
@wobba @copilot - think this is ready for review!

I've produced all the edge case file sizes and have uploaded them to SharePoint to see the file sizes in the document library. Then have pointed a search results webpart at the library and had my formatSPFileSize helper converting OWS_FILESIZEDISPLAY via a RefinableInt to a File Size display i.e. 873 bytes, 1 KB, 58.9KB, 1 GB, 41.3 GB etc

My Search experience now matches by document library for file size display.

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 4]
```diff
@@ -232,6 +232,59 @@ function helperHexToString(hex: any): string {
...
+        // a ~1018.88 KB file (1043333 B) shows as "0.99 MB", not "1019 KB".
+        const units = ['KB', 'MB', 'GB', 'TB'];
+        let value = b / 1024;
+        let i = 0;
+        while (value >= 1000 && i < units.length - 1) {
+            value /= 1024;
+            i++;
+        }
+        return formatSizeUnit(value, units[i]);
+    };
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts:285]: Dismissing — this is intentional and verified against live SharePoint. The promotion threshold is the **raw** value ≥ 1000, not the rounded display. A 999.7 KB file (`1023693 B`) renders as `1000 KB` in a real SharePoint library — promoting it to `0.98 MB` (Copilot's suggestion) would *break* that parity. Confirmed against a 90-row dataset of actual SharePoint values; `1000 <unit>` is correct SharePoint output, not a bug.

File: docs/extensibility/templating.md [hunk 5]
```diff
@@ -259,6 +259,44 @@ Supported output format tokens include all [dayjs format tokens](https://day.js.
...
+| 0 | (blank) |
+| 500 | `500 bytes` |
+| 1023 | `1023 bytes` |
+| 1024 | `1 KB` |
+| 1048575 | `1.00 MB` |
+| 1048576 | `1 MB` |
+| 10485760 | `10 MB` |
+| 1073741823 | `1.00 GB` |
+| 1073741824 | `1 GB` |
+| 44398312377 | `41.3 GB` |
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply docs/extensibility/templating.md:298]: Dimissed - Verified against the raw file — every row uses a single leading pipe (| … | … |) with a |---|---| separator. This is standard GFM and renders correctly on the docs site, consistent with the other tables in this file. No || present; false positive.

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 6]
```diff
@@ -232,6 +232,59 @@ function helperHexToString(hex: any): string {
...
+        if (Number.isInteger(value)) return `${value} ${unit}`;
+
+        let decimals = 0;
+        if (value < 10) decimals = 2;
+        else if (value < 100) decimals = 1;
+
+        return decimals > 0
+            ? `${value.toFixed(decimals)} ${unit}`
+            : `${Math.round(value)} ${unit}`;
+    }
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts:253]: Intentional and verified against live SharePoint. The promotion threshold is the <strong>raw</strong> derived value ≥ 1000, not the rounded display. A 999.7 KB file (<code>1023693 B</code>) renders as <code>1000 KB</code> in a real SharePoint library — promoting it to MB (as suggested) would <em>break</em> parity. Confirmed against a 90-row dataset of actual SharePoint values; <code>1000 KB</code> is correct output, not a bug.

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 7]
```diff
@@ -232,6 +232,59 @@ function helperHexToString(hex: any): string {
     return StringHelper.hexToString(String(hex));
 }
 
+const helperFormatSPFileSize: (bytes: any) => string = (() => {
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts:235]: Duplicate - Intentional and verified against live SharePoint. The promotion threshold is the raw derived value ≥ 1000, not the rounded display. A 999.7 KB file (1023693 B) renders as 1000 KB in a real SharePoint library — promoting it to MB (as suggested) would break parity. Confirmed against a 90-row dataset of actual SharePoint values; 1000 KB is correct output, not a bug.

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 8]
```diff
@@ -232,6 +232,59 @@ function helperHexToString(hex: any): string {
...
+            : `${Math.round(value)} ${unit}`;
+    }
+
+    /**
+     * Returns the byte count, or undefined (→ blank) for empty, non-numeric,
+     * negative, or non-integer input. Values above Number.MAX_SAFE_INTEGER
+     * (~8 PB) are also rejected; SharePoint's max file size is 250 GB, so this
+     * bound is unreachable in practice.
+     */
+    function tryGetValidFileSizeBytes(bytes: any): number | undefined {
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts:261]: Consistent with every other helper in this file, all typed <code>any</code>. The helper map is <code>Record&lt;string, (...args: any[]) =&gt; any&gt;</code> and Handlebars dispatches dynamically from untyped templates, so a stricter signature gives no real compile-time safety on the actual call path. Input is validated at runtime in <code>tryGetValidFileSizeBytes</code>.

File: search-parts/src/helpers/HandlebarsHelpers.ts [hunk 9]
```diff
@@ -232,6 +232,59 @@ function helperHexToString(hex: any): string {
...
+        const raw = typeof bytes === 'string' ? bytes.trim() : bytes;
+        if (raw === undefined || raw === '') return undefined;
+        if (!isNumber(raw)) return undefined;
+        const n = Number(raw);
+        if (!Number.isSafeInteger(n) || n < 0) return undefined;
+        return n;
+    }
+
+    return function helperFormatSPFileSize(bytes: any): string {
```
[the remark this answers is not in the document]
Comment (LeonArmston) [reply search-parts/src/helpers/HandlebarsHelpers.ts:270]: Consistent with every other helper in this file, all typed <code>any</code>. The helper map is <code>Record&lt;string, (...args: any[]) =&gt; any&gt;</code> and Handlebars dispatches dynamically from untyped templates, so a stricter signature gives no real compile-time safety on the actual call path. Input is validated at runtime in <code>tryGetValidFileSizeBytes</code>.

**Comment by LeonArmston** [comment]
Hi @wobba 

This PR is now ready for review.

A heads-up on the commit history: you may have seen this PR balloon to 40+ commits 😀 -  that was me working through GitHub Copilot's automated review suggestions via the web UI, which created a commit per edit. I've since squashed everything down to a single clean commit; the final diff is unchanged, just tidied up for review.

What this adds: a formatSPFileSize Handlebars helper that formats a raw byte count into a SharePoint-style file size (bytes / KB / MB / GB / TB), for use in templates via a slot mapped from a file-size managed property.

On the Copilot comments: I've gone through all of them, replied with rationale, and resolved each thread. One I'd specifically like to flag, since it's the one most likely to look like a bug at a glance: Copilot suggested adding "post-rounding promotion" so the helper never displays 1000 KB. I deliberately did not make that change — SharePoint genuinely renders 1000 KB for values like 999.7 KB, and promoting it would break parity with native SharePoint output. I verified this (and all the tier/rounding/boundary behaviour) empirically by uploading ~90 test files of known byte sizes to a real SharePoint library and confirming the helper's output matches exactly — including the sub-10 two-decimal cases, the 1000-byte promotion threshold, and the unit boundaries. I've added a code comment at that spot explaining why it's intentional. Happy to share the test dataset if useful.

The remaining Copilot notes (any vs unknown typing; a false positive about || table syntax) are also resolved with rationale inline — short version: the typing matches every other helper in the file, and the docs table uses standard single-pipe syntax that renders correctly.

Let me know if you'd like any changes. Thanks for taking a look!

**Comment by wobba** [comment]
Thank you.
